### PR TITLE
Add TektonListener to accept events and trigger pipeline/task runs

### PR DESCRIPTION
**Pull request by iancoffey**
# Changes

Proposal: https://docs.google.com/document/d/1tfG6dkARP2xko-5trDV3D62DJZiXQBZgLtLH7GyUopM/edit#

This is a mostly-functioning WIP of the proposal linked above. It adds a new type, PipelineListener and CRD which allow for a workflow to accept events and trigger pipeline runs end-to-end from any source. This implementation is meant to allow the project to accept any event type, with initial support for CloudEvents, and the ability to add event handling as needed for other types/sources.

## Related 
This concept started off in the build project, there is some good background there as well.

https://github.com/knative/build/issues/577
https://github.com/knative/build/pull/591

## TODO
 - [ ] Add tests
 - [ ] investigate ftest situation
 - [ ] end-to-end checksuite event success

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

```
Adds a new CRD and type `PipelineListener` to allow for PipelineRuns to be triggered automatically from events.
```

### CODE REVIEW

File: cmd/pipelinelistener/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      app: test-build-pipeline-listener
+  serviceName: test-build-pipeline-listener
+  template:
+    metadata:
+      labels:
+        role: test-build-pipeline-listener
+    spec:
+      serviceAccountName: tekton-pipelines-controller
+  listener-image: github.com/tektoncd/pipeline/cmd/pipelinelistener
+  event-type: com.github.checksuite
```
Comment (dlorenc) [inline cmd/pipelinelistener/README.md]: Is there a way to take data out of the event payload and thread that through into the pipeline via creation of a new resource, or similar?

File: cmd/pipelinelistener/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    matchLabels:
+      app: test-build-pipeline-listener
+  serviceName: test-build-pipeline-listener
+  template:
+    metadata:
+      labels:
+        role: test-build-pipeline-listener
+    spec:
+      serviceAccountName: tekton-pipelines-controller
+  listener-image: github.com/tektoncd/pipeline/cmd/pipelinelistener
```
Comment (dlorenc) [inline cmd/pipelinelistener/README.md]: What about storing info like the image used in a configmap or a flag somewhere?

File: cmd/pipelinelistener/README.md (hunk 1 above)
Comment (dlorenc) [reply cmd/pipelinelistener/README.md]: Here's a strawperson I put together on how binding eventdata into a resource might work: https://gist.github.com/dlorenc/bdd3dc952cd9b9cb10b837224b031214

Any thoughts?

Comment (iancoffey) [reply cmd/pipelinelistener/README.md]: I find this very interesting! Admittedly though, I am still processing how the example `EventBinding` type fits into the puzzle, inre: this PR.

Would that kind be used to swawn a listener behind the scenes based on the binding type? Or am I way off :)

File: cmd/pipelinelistener/README.md (hunk 2 above)
Comment (iancoffey) [reply cmd/pipelinelistener/README.md]: That sounds like a better plan, Ill make that change.

**Comment by khrm** [comment]
@iancoffey 

1. Doesn't this implementation create a stateful set for every repo?
2. Shouldn't this pr be raised wrt https://github.com/tektoncd/experimental/tree/master/webhooks-extension ?

3. Great idea of using runSpec to define pipelineRun . I have done an implementation which uses location of pipelinerun.yaml to define it.

**Comment by iancoffey** [comment]
@khrm hello, some answers inline:
> Doesn't this implementation create a stateful set for every repo?

This approach creates a listener for any run spec provided - there could also be multiple Listeners for one repo. Thinking of use cases - there may be different events that need to be handled, or you may build different parts of your monorepo based on different CI events. I think the implementation allows for almost any combo of repo/branch/pipeline but to be fair I have only tested 1:1:1.

> Shouldn't this pr be raised wrt https://github.com/tektoncd/experimental/tree/master/webhooks-extension ?

Yes, its possible. My goal for next Tuesdays Tekton WG meet is to understand where this should be added to the experimental repo or if this can land right here once feedback is addressed.

> Great idea of using runSpec to define pipelineRun . I have done an implementation which uses location of pipelinerun.yaml to define it.

Cool! Thanks. There are many updates coming via the proposal document that will likely change much of this but that part of the spec will likely stay and grow.

**Comment by bobcatfish** [comment]
@iancoffey can we close this PR now that we've got https://github.com/tektoncd/experimental/tree/master/tekton-listener ? :D

**Comment by iancoffey** [comment]
yes and ty!

**Pull request closed without merging**